Added PROXY Protocol TLV support

dobermai opened the pull request:
Motivation:

The current PROXY protocol implementation does not have support for optional Type-Length-Value fields. This pull requests adds the TLV values as specified in the PROXY protocol specification (http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt) and adds support for arbitrary TLVs.

Modifications:

The existing HAProxyMessage implements an additional TLV reading operation. A small bug in the AF_UNIX reader which didn’t set the reader index correctly was also fixed.

Result:

The PROXY protocol supports TLVs

netkins commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -73,6 +78,18 @@ private HAProxyMessage(
             String sourceAddress, String destinationAddress, int sourcePort, int destinationPort) {
 
+        this(protocolVersion, command, proxiedProtocol,
+             sourceAddress, destinationAddress, sourcePort, destinationPort, Collections.EMPTY_LIST);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace "Collections.EMPTY_LIST" by "Collections.emptyList()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1596)

normanmaurer replied:
@dobermai please address

netkins commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -73,6 +78,18 @@ private HAProxyMessage(
+    /**
+     * Creates a new instance
+     */
+    private HAProxyMessage(
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Constructor has 8 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS00107)

dobermai replied:
What is the preferred Netty way for avoid this netkins issue? Should a builder pattern be introduced? for the HAProxyMessage?

normanmaurer replied:
@dobermai ignore...

normanmaurer commented:
@trustin PTAL

> Am 29.09.2016 um 06:35 schrieb Obermaier Dominik <notifications@github.com>:
> 
> @dobermai commented on this pull request.
> 
> In codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
> 
> > @@ -73,6 +78,18 @@ private HAProxyMessage(
>              HAProxyProtocolVersion protocolVersion, HAProxyCommand command, HAProxyProxiedProtocol proxiedProtocol,
>              String sourceAddress, String destinationAddress, int sourcePort, int destinationPort) {
>  
> +        this(protocolVersion, command, proxiedProtocol,
> +             sourceAddress, destinationAddress, sourcePort, destinationPort, Collections.EMPTY_LIST);
> +    }
> +
> +    /**
> +     * Creates a new instance
> +     */
> +    private HAProxyMessage(
> What is the preferred Netty way for avoid this netkins issue? Should a builder pattern be introduced? for the HAProxyMessage?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -211,15 +233,94 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
+            }
+
+            return new HAProxySSLTLV(verify, client, encapsulatedTlvs, rawContent);
+        //If we're not dealing with a SSL Type, we can use the same mechanism
```
`//If` -> `// If`

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -211,15 +233,94 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
+
+    private static byte[] copyByteBufToByteArray(final ByteBuf header, final int length) {
+        final byte[] content = new byte[length];
+        header.readBytes(content);
```
use `header.getBytes(header.readerIndex, content);` so you not need to duplicate the `ByteBuf`

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -211,15 +233,94 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
+
+        case PP2_TYPE_SSL:
+
+            final byte[] rawContent = copyByteBufToByteArray(header.duplicate(), length);
```
see below... not need to duplicate()

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -211,15 +233,94 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
-     * @return                           {@link HAProxyMessage} instance
-     * @throws HAProxyProtocolException  if any portion of the header is invalid
+     * @param header a version 1 proxy protocol header
+     *
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -211,15 +233,94 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
+     * @param header a version 1 proxy protocol header
+     *
+     * @return {@link HAProxyMessage} instance
+     *
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -304,9 +406,11 @@ private static String ipBytestoString(ByteBuf header, int addressLen) {
-     * @return                           port as an integer
-     * @throws HAProxyProtocolException  if port is not a valid integer
+     * @param value the port
+     *
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -304,9 +406,11 @@ private static String ipBytestoString(ByteBuf header, int addressLen) {
+     * @param value the port
+     *
+     * @return port as an integer
+     *
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -326,50 +430,52 @@ private static int portStringToInt(String value) {
+            }
+            return;
+        case AF_UNIX:
+            return;
```
@dobermai I wonder if you can revert all the formatting changes and only include the "real" code changes in the PR.

dobermai replied:
@normanmaurer: Sure. The changes were applied after I used the Netty code template for IntelliJ since the original formatting seems to have an alternative coding style. Would be no problem to revert the formatting changes

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -326,50 +430,52 @@ private static int portStringToInt(String value) {
-     * @param port                       the UDP/TCP port
-     * @throws HAProxyProtocolException  if the port is out of range (0-65535 inclusive)
+     * @param port the UDP/TCP port
+     *
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -425,4 +531,13 @@ public int sourcePort() {
+     * TLVs are only available for the Proxy Protocol V2
+     */
+    public List<HAProxyTLV> getTlvs() {
+        return Collections.unmodifiableList(tlvs);
```
Why not store the `tlvs` already as immutable ?

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+ * Represents a {@link HAProxyTLV} of the type {@link io.netty.handler.codec.haproxy.HAProxyTLV.Type#PP2_TYPE_SSL}.
+ * This TLV encapsulates other TLVs and has additional information like verification information and a client bitfield.
+ */
+public class HAProxySSLTLV extends HAProxyTLV {
```
final

dobermai replied:
Is there a specific reason for making this class final? I can certainly make it final, I am just wondering what the rationale would be in this specific case.

normanmaurer replied:
we tend to make things final to make it easier to change things later. That that this has a package private constructor so it should be fine.

dobermai replied:
Makes sense. I made the class final, if there is no significant update in the PROXY protocol spec, it doesn't make any sense to create a subclass anyway

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.Set;
+
+/**
+ * Represents a {@link HAProxyTLV} of the type {@link io.netty.handler.codec.haproxy.HAProxyTLV.Type#PP2_TYPE_SSL}.
```
No need to use `fqcn` in the javadocs.

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+        super(Type.PP2_TYPE_SSL, (byte) 0x20, rawContent);
+
+        if (tlvs == null) {
+            throw new NullPointerException("tlvs");
```
`ObjectUtils.checkNonNull(...)`

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+        //Now parse the bitmask
+        clients = new HashSet<CLIENT>();
+        if ((clientBitField & 0x1) == 0x1) {
+
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+        this.verify = verify;
+        this.tlvs = tlvs;
+
+        //Now parse the bitmask
```
`//Now` -> `// Now`

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+            clients.add(CLIENT.PP2_CLIENT_CERT_CONN);
+        }
+        if ((clientBitField & 0x4) == 0x4) {
+
```
remove empty line

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+        if ((clientBitField & 0x2) == 0x2) {
+            clients.add(CLIENT.PP2_CLIENT_CERT_CONN);
+        }
+        if ((clientBitField & 0x4) == 0x4) {
```
shouldn't this be `!= 0`? Same for the bitmap stuff above

dobermai replied:
Yep you're right, that would be more readable. I'm adding an additional unit test to ensure nothing is broken :)

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    public List<HAProxyTLV> getEncapsulatedTLVs() {
+        return Collections.unmodifiableList(tlvs);
+    }
```
Usually we not use `get*` prefix. Please remove it

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+
+    @Override
+    public boolean encapsulatesOtherTLVs() {
+        //This TLV always encapsulates additional TLVs
```
Just add this to the java docs of it.

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,122 @@
+         *
+         * @return the {@link Type} of a TLV
+         */
+        public static Type getType(final byte byteValue) {
```
Please remove `get` prefix

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,122 @@
+        }
+        if (content == null) {
+            throw new NullPointerException("content");
+        }
```
Use `ObjectUtils.checkNotNull(...)`

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,122 @@
+     * Returns the raw content of the TLV as byte array
+     */
+    public byte[] getContent() {
+        return content;
```
This is not safe... use `content.clone()`

normanmaurer commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * Returns the {@link Type} of this TLV
+     */
+    public Type getType() {
```
Please remove all `get` prefixes

normanmaurer commented on codec-haproxy/src/test/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoderTest.java:
```diff
@@ -587,20 +590,69 @@ public void testV2UnknownProtocolDecode() {
+
+        final HAProxyTLV thirdTLV = tlvs.get(2);
+        assertEquals(HAProxyTLV.Type.PP2_TYPE_SSL_CN, thirdTLV.getType());
+        assertArrayEquals("LEAF".getBytes(), thirdTLV.getContent());
```
`getBytes(CharsetUtil.US_ASCII)`

dobermai commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -182,26 +202,28 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
                 addressLen = addressEnd - startIdx;
             }
             dstAddress = header.toString(startIdx, addressLen, CharsetUtil.US_ASCII);
+            //AF_UNIX defines that excactly 108 bytes are reserved for the address
```
unfortunately the reader index is not updated before (the toString method doesn't modify the reader index and the forEachByte method also doesn't modify the reader index. If we don't update the index, we're starting to read on the wrong place for the TLVs

normanmaurer replied:
@dobermai this sounds like a bug then... Can you open another PR where you just fix this ?

dobermai replied:
Sure, will create a PR.

dobermai replied:
@normanmaurer: Created a separate PR for this: #5883

dobermai commented:
@normanmaurer: Thanks for reviewing! I addressed all your comments except the one with the formatting. I used the Netty IntelliJ template from http://netty.io/wiki/developer-guide.html. Is there a more current one I should use? Seems the original class I modified did not follow the IntelliJ template. Let me know if you'd like me to re-format to the "old" format.

normanmaurer commented:
@dobermai can you please address my comments and rebase ?

dobermai commented:
@normanmaurer: All comments are addressed and the code is rebased against current 4.1 branch

normanmaurer commented:
@dobermai did another round of review. Please check

dobermai commented:
@normanmaurer: Sorry, seems I'm unable to find these additional reviews in the Github interface. The interface shows me that there are comments 12 days ago, which I have addressed (added PR #5883 and made HAProxySSLTLV final). Are there additional comments I have missed?

normanmaurer commented:
@dobermai doh! Forgot to click "submit" :(

dobermai commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * The possible values of the clientfield bitmask
+     */
+    public enum CLIENT {
```
I checked the Netty Source and most of the classes use upper-case enum values, only a few classes use lower case enum values. Since the HAProxy protocol also defines these enum values in upper case (see http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt), I believe it would be beneficial to have the values in upper case. If camel case enum values are mandatory I can change this of course.

dobermai replied:
done :)

dobermai commented:
@normanmaurer: Addressed your comments. Let me know if you want me to squash the commits or if you have any additional comments I should address.

normanmaurer commented:
@trustin could you have a look as well ?

normanmaurer commented:
@trustin PTAL ...

normanmaurer commented:
@Scottmitch PTAL as well.

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,106 @@
+     * @param tlvs the encapsulated {@link HAProxyTLV}s
+     * @param rawContent the raw TLV content
+     */
+    HAProxySSLTLV(final int verify, final byte clientBitField, final List<HAProxyTLV> tlvs, final byte[] rawContent) {
```
is the `verify` terminology defined in a specification somewhere? If not can we provide a bit more descriptive name (e.g. `verificationResult`)? If it is can you provide a link in the javadocs?

dobermai replied:
This field is actually called "verify" by the spec, there is no further explanation in the spec, though. I will add a Javadoc Link

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,106 @@
+    HAProxySSLTLV(final int verify, final byte clientBitField, final List<HAProxyTLV> tlvs, final byte[] rawContent) {
+        super(Type.PP2_TYPE_SSL, (byte) 0x20, rawContent);
+
+        checkNotNull(tlvs, "tlvs");
```
this will be done by `Collections.unmodifiableList` below ... consider removing

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,106 @@
+     * Returns an unmodifiable Set of {@link Client} values for this SSL TLV
+     */
+    public Set<Client> clients() {
+        return Collections.unmodifiableSet(clients);
```
`clients` does not appear to be modified after construction ... can we just wrap the member variable in a `Collections.unmodifiableSet` in the constructor after the set is built?

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,120 @@
+     * Returns <code>true</code> if this TLV encapsulates another TLV (this is typically only the case
+     * for {@link HAProxySSLTLV}s
+     */
+    public boolean encapsulatesOtherTLVs() {
```
It doesn't look like this is used ... what is the intended use case for this? in `HAProxyMessage.java` an `instanceof HAProxySSLTLV` check is done to access the `encapsulatedTLVs()` method which presumably is intended to be related to this method? It seems like to get access to what this field determines ... you will have to cast to some other type anyways so I don't understand the value of this method.

Scottmitch commented on codec-haproxy/src/test/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoderTest.java:
```diff
@@ -587,20 +590,71 @@ public void testV2UnknownProtocolDecode() {
+        assertEquals(3, tlvs.size());
+        final HAProxyTLV firstTlv = tlvs.get(0);
+        assertEquals(HAProxyTLV.Type.PP2_TYPE_SSL, firstTlv.type());
+        assertTrue(firstTlv instanceof HAProxySSLTLV);
```
remove this line and do the cast directly and you will get the same result (`ClassCastException`)

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,106 @@
+            clients = Collections.emptySet();
+        } else {
+            // Now parse the bitmask
+            clients = new HashSet<Client>(4);
```
instead of maintaining a set for this can we just keep a `byte` (or `int`) bit mask and expose accessor methods on the class like `boolean isPP2ClientCertConn()` etc..? This has the potential to save a bunch of allocation/deallocation.

dobermai commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -214,7 +232,99 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
+        final int length = header.readUnsignedShort();
+        switch (type) {
+        case PP2_TYPE_SSL:
+            final byte[] rawContent = copyByteBufToByteArray(header, length);
```
The intention was that the HAProxyTLV may be used by business logic and thus I see two reasons for the byte array: 

1. The can reside in memory where it's outside of Netty's control. This has the advantage that no reference counting needs to be in place and the TLV values can be in memory as long as needed by the business application (in our use case the SSL information is actually in memory all the time for each channel!)
2. While the ByteBuf API is incredibly useful for the networking layer, it makes consuming "plain ol'" bytes harder for the business layer (which in our case expects byte arrays), because of the distinction between on- and off-heap memory. So the main argument here would be usability at the cost of a copy operation.

If the copy operation would be a show-stopper, I can certainly use ByteBufs here instead

normanmaurer replied:
@dobermai please extend `ByteBufHolder` ... If a user needs a byte[] its possible to do a memory copy later on.

dobermai commented:
@normanmaurer @Scottmitch thanks for reviewing! Addressed all the comments and pushed. Let me know if I should rebase and squash

normanmaurer commented:
@dobermai ping... please address our last comments so we can merge this

dobermai commented:
@normanmaurer: Sorry for not coming back to you earlier. I'll try to schedule these changes in this week.

normanmaurer commented:
No worries.... thanks

> Am 27.11.2016 um 22:55 schrieb Obermaier Dominik <notifications@github.com>:
> 
> @normanmaurer: Sorry for not coming back to you earlier. I'll try to schedule these changes in this week.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

dobermai commented:
@normanmaurer: Ok, now a ByteBufHolder is used. I'm not 100% sure if the retain() call in HAProxyMessage Line 249 is the most elegant solution here. If you see a better way, please let me know.

Ping me if you want me to rebase and squash.

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    public HAProxyTLV copy() {
+        return (HAProxyTLV) super.copy();
```
casting to `HAProxyTLV` will cause a ClassCastException. Instead do this:

```java
return new HAProxyTLV(type, typeByteValue, super.copy());
```

same comment for the other methods in this class which do the same thing.

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyTLV.java:
```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    public HAProxyTLV duplicate() {
+        return (HAProxyTLV) super.duplicate();
```
Related to the previous comment ... The typical way this is done in other classes is something like this:

```java
    @Override
    public HAProxyTLV duplicate() {
        return replace(content().duplicate());
    }

    @Override
    public HAProxyTLV replace(ByteBuf content) {
        return new HAProxyTLV(type, typeByteValue, content);
    }
```

Then all of the other methods which do the same thing (copy, retainedDuplicate, etc...) can be written the same way.

Scottmitch commented on codec-haproxy/src/test/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoderTest.java:
```diff
@@ -587,20 +589,76 @@ public void testV2UnknownProtocolDecode() {
+        ch.writeInbound(copiedBuffer(bytes));
+        Object msgObj = ch.readInbound();
+        assertEquals(startChannels - 1, ch.pipeline().names().size());
+        assertTrue(msgObj instanceof HAProxyMessage);
```
consider just removing this check ... a ClassCastException will be thrown on the next line anyways.

Scottmitch commented on codec-haproxy/src/test/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoderTest.java:
```diff
@@ -587,20 +589,76 @@ public void testV2UnknownProtocolDecode() {
+        assertArrayEquals("LEAF".getBytes(CharsetUtil.US_ASCII), thirdContent);
+
+        assertEquals(true, sslTlv.encapsulatedTLVs().contains(secondTlv));
+        assertEquals(true, sslTlv.encapsulatedTLVs().contains(thirdTLV));
```
all of the `HAProxyTLV` need to released at some point in this test and they are not used passed this point so it should be safe to release after this line.

Scottmitch commented on codec-haproxy/src/test/java/io/netty/handler/codec/haproxy/HAProxySSLTLVTest.java:
```diff
@@ -0,0 +1,62 @@
+
+        assertEquals(false, noClientTlv.isPP2ClientCertConn());
+        assertEquals(false, noClientTlv.isPP2ClientSSL());
+        assertEquals(false, noClientTlv.isPP2ClientCertSess());
```
all the `HAProxySSLTLV` objects must be released

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxySSLTLV.java:
```diff
@@ -0,0 +1,86 @@
+    public List<HAProxyTLV> encapsulatedTLVs() {
+        return tlvs;
+    }
+
```
nit: kill line

Scottmitch commented on codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java:
```diff
@@ -214,7 +232,71 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
+        List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>(4);
+
+        do {
+            haProxyTLVs.add(haProxyTLV.retain());
```
it seems strange that we only `retain()` the "top level" `HAProxyTLV` objects. For example wouldn't you also expect all the `((HAProxySSLTLV) haProxyTLV).encapsulatedTLVs()` added below to be released too? The call to `encapsulatedTlvs.add(haProxyTLV);` below doesn't account for `retain()` either. It may be more strait forward to do a retain (e.g `readRetainedSlice()` or `retainedSlice()`) in `readNextTLV` when ever we get the "raw content" (for `HAProxySSLTLV` and `HAProxyTLV`)

dobermai commented:
Thanks for the thorough review. Seems I misunderstood something about the ByteBufHolder. I try to fix these things as quick as possible, hopefully this or next week.

normanmaurer commented:
@Scottmitch can you check again please so we get this merged ?

Scottmitch commented:
@normanmaurer - I believe this has been overtaken by https://github.com/netty/netty/pull/6245 ... I will close this for now.
